File: matrix-multiplication/matrix_multiplication.go
package main
import "fmt"

func matrixMultiply(a, b [][]int) [][]int {
    rowsA, colsA := len(a), len(a[0])
    colsB := len(b[0])
    
    result := make([][]int, rowsA)
    for i := range result {
        result[i] = make([]int, colsB)
    }
    
    for i := 0; i < rowsA; i++ {
        for j := 0; j < colsB; j++ {
            for k := 0; k < colsA; k++ {
                result[i][j] += a[i][k] * b[k][j]
            }
        }
    }
    
    return result
}

func addMatrices(a, b [][]int) [][]int {
    rows, cols := len(a), len(a[0])
    result := make([][]int, rows)
    for i := range result {
        result[i] = make([]int, cols)
        for j := range result[i] {
            result[i][j] = a[i][j] + b[i][j]
        }
    }
    return result
}

func subtractMatrices(a, b [][]int) [][]int {
    rows, cols := len(a), len(a[0])
    result := make([][]int, rows)
    for i := range result {
        result[i] = make([]int, cols)
        for j := range result[i] {
            result[i][j] = a[i][j] - b[i][j]
        }
    }
    return result
}

func printMatrix(matrix [][]int) {
    for _, row := range matrix {
        for _, val := range row {
            fmt.Printf("%d ", val)
        }
        fmt.Println()
    }
}

func main() {
    a := [][]int{{1, 2}, {3, 4}}
    b := [][]int{{5, 6}, {7, 8}}
    
    result := matrixMultiply(a, b)
    fmt.Println("Result:")
    printMatrix(result)
}

File: matrix-multiplication/matrix_multiplication.c
#include <stdio.h>
#include <stdlib.h>

void matrixMultiply(int **a, int **b, int **result, int rowsA, int colsA, int colsB) {
    for (int i = 0; i < rowsA; i++) {
        for (int j = 0; j < colsB; j++) {
            result[i][j] = 0;
            for (int k = 0; k < colsA; k++) {
                result[i][j] += a[i][k] * b[k][j];
            }
        }
    }
}

void printMatrix(int **matrix, int rows, int cols) {
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            printf("%d ", matrix[i][j]);
        }
        printf("\n");
    }
}

int main() {
    int rowsA = 2, colsA = 2, colsB = 2;
    
    int **a = malloc(rowsA * sizeof(int*));
    int **b = malloc(colsA * sizeof(int*));
    int **result = malloc(rowsA * sizeof(int*));
    
    for (int i = 0; i < rowsA; i++) {
        a[i] = malloc(colsA * sizeof(int));
        result[i] = malloc(colsB * sizeof(int));
    }
    for (int i = 0; i < colsA; i++) {
        b[i] = malloc(colsB * sizeof(int));
    }
    
    // Initialize matrices
    a[0][0] = 1; a[0][1] = 2;
    a[1][0] = 3; a[1][1] = 4;
    
    b[0][0] = 5; b[0][1] = 6;
    b[1][0] = 7; b[1][1] = 8;
    
    matrixMultiply(a, b, result, rowsA, colsA, colsB);
    printMatrix(result, rowsA, colsB);
    
    // Free memory
    for (int i = 0; i < rowsA; i++) {
        free(a[i]);
        free(result[i]);
    }
    for (int i = 0; i < colsA; i++) {
        free(b[i]);
    }
    free(a); free(b); free(result);
    
    return 0;
}